src/dmd/backend/dwarfdbginf.d

```diff
@@ -1277,40 +1309,92 @@ static if (1)
         }
     }
 
+    /*************************************
+     * Add a directory to `lineDirectories`
+     */
+    uint dwarf_line_add_directory(const(char)* path)
+    {
+        assert(path);
+        return dwarf_line_add_directory(path[0 .. strlen(path)]);
+    }
+
+    /// Ditto
+    extern(D) uint dwarf_line_add_directory(const(char)[] path)
+    {
+        return addToAAchars(lineDirectories, retrieveDirectory(path));
+    }
 
     /*************************************
-     * Add a file to the .debug_line header
+     * Add a file to `infoFileName_table`
      */
-    int dwarf_line_addfile(const(char)* filename)
+    uint dwarf_line_addfile(const(char)* path)
     {
-        return dwarf_line_addfile(filename ? filename[0 .. strlen(filename)] : null);
+        assert(path);
+        return dwarf_line_addfile(path[0 .. strlen(path)]);
     }
 
     /// Ditto
-    extern(D) int dwarf_line_addfile(const(char)[] filename)
+    extern(D) uint dwarf_line_addfile(const(char)[] path)
     {
-        if (!infoFileName_table)
+        return addToAAchars(infoFileName_table, path);
+    }
+
+    /*************************************
+     * Adds `str` to `aachars`, and assigns a new index if none
+     *
+     * Params:
+     *      aachars = AAchars where to add `str`
+     *      str = string to add to `aachars`
+     */
+    extern(D) uint addToAAchars(ref AAchars* aachars, const(char)[] str)
+    {
+        if (!aachars)
         {
-            infoFileName_table = AAchars.create();
-            linebuf_filetab_end = debug_line.buf.length();
+            aachars = AAchars.create();
         }
 
-        uint *pidx = infoFileName_table.get(filename);
+        uint *pidx = aachars.get(str);
         if (!*pidx)                 // if no idx assigned yet
         {
-            *pidx = infoFileName_table.length(); // assign newly computed idx
-
-            size_t before = debug_line.buf.length();
-            debug_line.buf.writeString(filename);
-            debug_line.buf.writeByte(0);      // directory table index
-            debug_line.buf.writeByte(0);      // mtime
-            debug_line.buf.writeByte(0);      // length
-            linebuf_filetab_end += debug_line.buf.length() - before;
+            *pidx = aachars.length(); // assign newly computed idx
         }
-
         return *pidx;
     }
 
+    /**
+     * Extracts the file name from `path`.
+     *
+     * Params:
+     *      path = Full path containing the filename and the directory
+     * Returns:
+     *      The file name
+     */
+    extern(D) const(char)[] retrieveFilename(const(char)[] path)
```

Comment:
This function duplicates the functionality of filespec.filespecname(), while failing on Windows. It also presumes a 0 terminated path, while returning a string that may not be 0 terminated.